### PR TITLE
Implement Indexing for RabitQ

### PR DESCRIPTION
**Pull request by tvhong**
### Notes
Implement the Indexing Phase for RabitQ.

Paper: https://arxiv.org/abs/2405.12497
My summary of the indexing phase: https://github.com/hicder/muopdb/issues/32#issuecomment-2456416853

Please note that the current implementation is very inefficient since I don't have much experience writing high performance code, but it should at least be correct.

### CODE REVIEW

**Review by tyb0807 — COMMENTED** [review]
Thanks! Just one high level question, why `Array1` and `Array2` instead of `Vec` and `Vec<Vec>`? Asking just for the sake of consistency with other parts of the code base.

**Comment by tvhong** [comment]
@tyb0807 The querying phase needs to do some matrix multiplication and vector dot products (see https://github.com/hicder/muopdb/issues/32#issuecomment-2466578455). So, I store these as ndarray objects.

That being said, I'm not set on the data type at the moment. Maybe `ndarray` doesn't support serialization well, or that we should use `nalgebra` since we're only dealing with low-dimensionality data. Once that happens, we can easily change the data type.

**Review by hicder — CHANGES_REQUESTED** [review]
you'll need to use the normalized vector

File: rs/quantization/src/rabitq.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    // The quantization code for each data point
+    // Notation: $\bar{x}_b$
+    // Dimension: $N \times D$
+    pub quantization_codes: Vec<BitVec>,
+
+    // The dot products of each quantized vector with its data point
+    // Notation: $<\bar{o}, o>$
+    // Dimension: $N$
+    pub quantized_vector_dot_products: Array1<f32>,
+}
```
[the remark this answers is not in the document]
Comment (tvhong) [reply rs/quantization/src/rabitq.rs]: I see. Running `cargo fmt` fixed this.

**Comment by tvhong** [comment]
Updated the code to use normalized vectors instead.

**Pull request merged**